File: section1/TwoSum.java
import java.util.Arrays;

class Solution {
    public int[] twoSum(int[] nums, int target) {

        int[][] arr = new int[nums.length][2];

        for (int i = 0; i < nums.length; i++) {
            arr[i][0] = i;
            arr[i][1] = nums[i];
        }

        Arrays.sort(arr, (o1, o2) -> o2[1] - o1[1]);

        int left = 0, right = nums.length - 1;
        while (left < right) {
            int sum = arr[left][1] + arr[right][1];

            if (sum == target) {
                return new int[]{ arr[left][0], arr[right][0] };
            } else if (sum < target) {
                left += 1;
            } else {
                right -= 1;
            }
        }

        return new int[]{ -1, -1 };
    }
}

File: section2/two-sum.py
class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        nums = [(i, n) for i, n in enumerate(nums)]
        # nums.sort(key=lambda num: num[1])
        sorted_nums = sorted(nums, key=lambda num: num[1])
        i, j = 0, len(sorted_nums) - 1

        while i < j:
            sum = sorted_nums[i][1] + sorted_nums[j][1]
            if sum == target:
                return [sorted_nums[i][0], sorted_nums[j][0]]
            elif sum > target:
                j -= 1
            elif sum < target:
                i += 1

        return None
